editor/src/components/canvas/dom-walker.ts

```diff
@@ -428,112 +249,227 @@ function mergeFragmentMetadata(
-      const refOfContainer = containerRef.current
-      if (ObserversAvailable && resizeObserver != null && mutationObserver != null) {
-        Array.from(document.querySelectorAll(`#${CanvasContainerID} *`)).map((elem) => {
-          resizeObserver.observe(elem)
-        })
-        mutationObserver.observe(refOfContainer, MutationObserverConfig)
-      }
+export function createDomWalkerMutableState(
+  editorStoreApi: UtopiaStoreAPI,
+): DomWalkerMutableStateData {
+  const mutableData: DomWalkerMutableStateData = {
+    invalidatedPaths: emptySet(),
+    invalidatedScenes: emptySet(),
+    invalidatedPathsForStylesheetCache: emptySet(),
+    initComplete: true,
+
+    mutationObserver: null as any,
+    resizeObserver: null as any,
+  }
+
+  const observers = initDomWalkerObservers(mutableData, editorStoreApi)
+  mutableData.mutationObserver = observers.mutationObserver
+  mutableData.resizeObserver = observers.resizeObserver
+
+  return mutableData
+}
+
+export const DomWalkerMutableStateCtx = React.createContext<DomWalkerMutableStateData | null>(null)
+function useDomWalkerMutableStateContext() {
+  return forceNotNull(
+    `DomWalkerMutableStateCtx needs a Provider`,
+    React.useContext(DomWalkerMutableStateCtx),
+  )
+}
+
+interface RunDomWalkerParams {
+  // from dom walker props
+  selectedViews: Array<ElementPath>
+  scale: number
+  additionalElementsToUpdate: Array<ElementPath>
+
+  domWalkerMutableState: DomWalkerMutableStateData
+  rootMetadataInStateRef: { readonly current: readonly ElementInstanceMetadata[] }
+}
 
-      // getCanvasRectangleFromElement is costly, so I made it lazy. we only need the value inside globalFrameForElement
-      const containerRect = lazyValue(() => {
-        return getCanvasRectangleFromElement(refOfContainer, props.scale)
+export function runDomWalker({
+  domWalkerMutableState,
+  selectedViews,
+  scale,
+  additionalElementsToUpdate,
+  rootMetadataInStateRef,
+}: RunDomWalkerParams): { metadata: ElementInstanceMetadata[]; cachedPaths: ElementPath[] } | null {
+  const needsWalk =
+    !domWalkerMutableState.initComplete ||
+    domWalkerMutableState.invalidatedPaths.size > 0 ||
+    domWalkerMutableState.invalidatedScenes.size > 0
+
+  if (!needsWalk) {
+    return null // early return to save performance
+  }
+
+  const LogDomWalkerPerformance =
+    isFeatureEnabled('Debug mode – Performance Marks') && PERFORMANCE_MARKS_ALLOWED
+
+  const canvasRootContainer = document.getElementById(CanvasContainerID)
+
+  if (canvasRootContainer != null) {
+    if (LogDomWalkerPerformance) {
+      performance.mark('DOM_WALKER_START')
+    }
+    // Get some base values relating to the div this component creates.
+    if (
+      ObserversAvailable &&
+      domWalkerMutableState.resizeObserver != null &&
+      domWalkerMutableState.mutationObserver != null
+    ) {
+      Array.from(document.querySelectorAll(`#${CanvasContainerID} *`)).map((elem) => {
```

Comment:
The `NodeList` that comes back from `querySelectorAll` has a `forEach` function which should simplify this.
```suggestion
      document.querySelectorAll(`#${CanvasContainerID} *`).forEach((elem) => {
```

Reply:
*cough*

Reply:
fixed!